### PR TITLE
Version cleanup

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
A lot of people are not going to like this

### CODE REVIEW

**Comment by SMillerDev** [comment]
@Homebrew/maintainers this is likely going to affect a lot of people. Not everything is officially EOL, some just didn't have a release in 5 years and are versioned.

**Comment by jonchang** [comment]
What do the analytics look like for these versioned formulae?

**Comment by SMillerDev** [comment]
Haven't looked at all of them but gcc and postgres is used pretty heavily. Not really willing to keep unsupported things around because people use it though.

**Comment by vitorgalvao** [comment]
<details>
<summary>Analytics</summary>

```
antlr@2
30 days: 289 (#1686)
90 days: 698 (#1812)
365 days: 1,792 (#2169)

autoconf@2.13
30 days: 215 (#1912)
90 days: 615 (#1919)
365 days: 2,491 (#1907)

boost@1.55
30 days: 86 (#2709)
90 days: 228 (#2796)
365 days: 1,001 (#2719)

boost@1.57
30 days: 45 (#3471)
90 days: 141 (#3343)
365 days: 880 (#2856)

boost@1.59
30 days: 72 (#2882)
90 days: 215 (#2847)
365 days: 2,090 (#2029)

boost@1.60
30 days: 488 (#1340)
90 days: 1,738 (#1200)
365 days: 6,061 (#1282)

elasticsearch@2.4
30 days: 352 (#1541)
90 days: 1,001 (#1544)
365 days: 4,653 (#1432)

elasticsearch@5.6
30 days: 730 (#1111)
90 days: 2,177 (#1089)
365 days: 11,296 (#932)

gcc@4.9
30 days: 996 (#954)
90 days: 2,953 (#939)
365 days: 13,668 (#847)

gcc@5
30 days: 1,008 (#950)
90 days: 2,367 (#1040)
365 days: 10,731 (#956)

gcc@6
30 days: 990 (#959)
90 days: 3,949 (#784)
365 days: 15,993 (#777)

go@1.10
30 days: 182 (#2039)
90 days: 523 (#2031)
365 days: 3,126 (#1721)

go@1.11
30 days: 224 (#1886)
90 days: 727 (#1780)
365 days: 7,366 (#1144)

go@1.12
30 days: 2,258 (#616)
90 days: 8,293 (#545)
365 days: 25,602 (#613)

go@1.9
30 days: 249 (#1800)
90 days: 707 (#1800)
365 days: 4,156 (#1515)

kibana@5.6
30 days: 116 (#2406)
90 days: 337 (#2397)
365 days: 1,771 (#2182)

maven@3.2
30 days: 500 (#1324)
90 days: 1,197 (#1429)
365 days: 3,880 (#1549)

maven@3.3
30 days: 608 (#1216)
90 days: 1,598 (#1255)
365 days: 5,115 (#1375)

maven@3.5
30 days: 1,246 (#852)
90 days: 3,187 (#891)
365 days: 11,244 (#934)

mysql@5.6
30 days: 6,629 (#358)
90 days: 21,314 (#326)
365 days: 81,906 (#323)

postgresql@9.4
30 days: 873 (#1026)
90 days: 2,482 (#1016)
365 days: 12,883 (#875)

protobuf@3.6
30 days: 316 (#1617)
90 days: 955 (#1576)
365 days: 4,878 (#1408)

redis@3.2
30 days: 511 (#1307)
90 days: 1,298 (#1376)
365 days: 5,488 (#1334)

v8@3.15
30 days: 2,507 (#591)
90 days: 7,468 (#578)
365 days: 15,944 (#781)
```
</details>

**Comment by jonchang** [comment]
Thanks @vitorgalvao! I've formatted it into a table:

formula | install (30) | rank (30) | install (90) | rank (90) | install (365) | rank (365) |
|-------|--------------|-----------|--------------|-----------|---------------|------------|
antlr@2 | 289 | 1686 | 698 | 1812| 1,792 | 2169
autoconf@2.13 | 215 | 1912 | 615 | 1919| 2,491 | 1907
boost@1.55 | 86 | 2709 | 228 | 2796| 1,001 | 2719
boost@1.57 | 45 | 3471 | 141 | 3343| 880 | 2856
boost@1.59 | 72 | 2882 | 215 | 2847| 2,090 | 2029
boost@1.60 | 488 | 1340 | 1,738 | 1200| 6,061 | 1282
elasticsearch@2.4 | 352 | 1541 | 1,001 | 1544| 4,653 | 1432
elasticsearch@5.6 | 730 | 1111 | 2,177 | 1089| 11,296 | 932
gcc@4.9 | 996 | 954 | 2,953 | 939| 13,668 | 847
gcc@5 | 1,008 | 950 | 2,367 | 1040| 10,731 | 956
gcc@6 | 990 | 959 | 3,949 | 784| 15,993 | 777
go@1.10 | 182 | 2039 | 523 | 2031| 3,126 | 1721
go@1.11 | 224 | 1886 | 727 | 1780| 7,366 | 1144
go@1.12 | 2,258 | 616 | 8,293 | 545| 25,602 | 613
go@1.9 | 249 | 1800 | 707 | 1800| 4,156 | 1515
kibana@5.6 | 116 | 2406 | 337 | 2397| 1,771 | 2182
maven@3.2 | 500 | 1324 | 1,197 | 1429| 3,880 | 1549
maven@3.3 | 608 | 1216 | 1,598 | 1255| 5,115 | 1375
maven@3.5 | 1,246 | 852 | 3,187 | 891| 11,244 | 934
mysql@5.6 | 6,629 | 358 | 21,314 | 326| 81,906 | 323
postgresql@9.4 | 873 | 1026 | 2,482 | 1016| 12,883 | 875
protobuf@3.6 | 316 | 1617 | 955 | 1576| 4,878 | 1408
redis@3.2 | 511 | 1307 | 1,298 | 1376| 5,488 | 1334
v8@3.15 | 2,507 | 591 | 7,468 | 578| 15,944 | 781

Linuxbrew/core needs gcc, I believe `gcc-5`, so that needs to be retained. I think we're hoping to change default compilers to a newer version "soon" though. cc @Homebrew/linux

**Comment by iMichka** [comment]
Yeah, we might switch to gcc 7 (when we move to ubuntu 18.04 as build system), but not right now.

You can delete these in my opinion. We will keep a copy around in linuxbrew-core for the time being.

**Comment by tschoonj** [comment]
`gjs` really needs to build against `autoconf@213`. If you insist on removing the formula, how about I turn it into a resource for `gjs` instead?

**Comment by Bo98** [comment]
@tschoonj That would make sense, unless a versioned formula would be of significant benefit. I see the install count is still somewhat high.

Do you know why it needs 2.13? Seems like something worth mentioning as a comment in the formula file, particularly since 2.13 is over 20 years old.

**Comment by MikeMcQuaid** [comment]
Thanks @vitorgalvao and @jonchang, that makes it much easier to follow.

It would be helpful to know for each of these what criteria they violate on https://docs.brew.sh/Versions.

> some just didn't have a release in 5 years and are versioned.

This, for example, I don't think is a good enough reason by itself unless it violates other things in https://docs.brew.sh/Versions. 

Remember: our (Homebrew's) goal is to be useful more than "clean". Unless formulae are causing us issues we should strive to keep them in Homebrew.

https://github.com/Homebrew/brew/issues/7029 would be a nicer fit to reduce the burden without needing the same level of cleanup.

**Comment by SMillerDev** [comment]
I'll check for the things without an official EOL if they have known security issues. The most used ones (go/pgsql) are officially EOL though.

**Comment by tschoonj** [comment]
> @tschoonj That would make sense, unless a versioned formula would be of significant benefit, which it seems it might be. I see the install count is still somewhat high.
> 
> Do you know why it needs 2.13? Seems like something worth mentioning as a comment in the formula file, particularly since 2.13 is over 20 years old.

The build system of SpiderMonkey (the Firefox Javascript engine) is very complicated and highly customised. My guess is that at some point they decided to put in a lot of Autoconf macros whose syntax was changed in later releases in an incompatible way and they never bothered to change them...

**Comment by sjackman** [comment]
I would like to see a deprecation DSL and policy implemented before we start removing large swaths of versioned formulae.

**Comment by SMillerDev** [comment]
I removed everything that was just old, all these versions are no longer supported by upstream.

**Comment by fxcoudert** [comment]
We've typically kept some older compilers because they're useful to bootstrap other stuff and some software (not in Homebrew, but in the wider ecosystem) can have strict expectations on a specific compiler.

I'd ask you to keep some of the go versions, and gcc ≥ 6 at least (which works on all macOS versions).

> We will keep a copy around in linuxbrew-core for the time being.

This is a sign that we should keep things.

**Comment by SMillerDev** [comment]
I've removed llvm from the removal list. No CVEs and linuxbrew still needs it, as for the GO releases, those are officially unsupported. What would be the motivation for keeping them?

**Comment by MikeMcQuaid** [comment]
> What would be the motivation for keeping them?

For me:
- people are still using them
- they aren't causing us any (GitHub) issues
- they still build on latest versions of macOS
- they aren't failing CI in other ways

**Comment by SMillerDev** [comment]
> Upstream should have a release branch for the versioned formulae version and still make security updates for that version, when necessary. For example, PHP 5.5 was not a supported version but PHP 7.2 was in January 2018.

It does not meet this requirement in https://docs.brew.sh/Versions though. I'm fine keeping them, just wanted to be clear on the official EOL. (https://golang.org/doc/devel/release.html#policy) for reference.

**Comment by MikeMcQuaid** [comment]
We have a 2.2.12 tag now so we can use the deprecation stuff.

**Comment by SMillerDev** [comment]
I found another document stating that mysql 5.6 isn't deprecated and all builds are passing so I think this is ready to go now. @Homebrew/core

**Comment by MikeMcQuaid** [comment]
Can we deprecate these instead of removing them (as discussed above)?

**Comment by SMillerDev** [comment]
With a date in the past then? I'm still not really sure what that would help

**Comment by MikeMcQuaid** [comment]
> With a date in the past then?

@smillerdev that or with no date. It'll help because it'll let us test out the deprecation feature and stop these formulae being tested in CI (which presumably is part of the reason you would like to have them removed).

I'd suggest perhaps doing:
```
disable! date: $SOME_DATE
```

as-is that'll be a no-op but after https://github.com/Homebrew/brew/pull/7462 it'll deprecate, too.

Alternatively:
```
deprecate!
disable! date: $SOME_DATE
```

will immediately deprecate and disable it later (and we can remove the `deprecate!` once https://github.com/Homebrew/brew/pull/7462 hits a tag).

**Comment by SMillerDev** [comment]
> which presumably is part of the reason you would like to have them removed

Nah, I just think people shouldn't use EOL software.

---

I'll set the depreciation dates to the EOL dates later today.

**Comment by MikeMcQuaid** [comment]
Also: looks like the `kops` upstream tarball SHA256 has changed.

**Comment by iMichka** [comment]
What I am missing is the "removal" date. We have a deprecation date, and a disable flag. But when do we remove the formula? Could we display a message that a formula will be removed 1 month / 6 months after it has been disabled?

**Comment by MikeMcQuaid** [comment]
> Could we display a message that a formula will be removed 1 month / 6 months after it has been disabled?

We could but I'm not sure committing to maintainers manually doing something on a given date is wise. We could have a GitHub Action that does that automatically, though 😂. I'm open to ideas regardless.

**Comment by Bo98** [comment]
> Also: looks like the `kops` upstream tarball SHA256 has changed.

Yeah, that's the thing that relied on git short hashes and GitHub is generating a different tarball now with a hash one character longer. I can confirm the checksum change isn't malicious.

**Comment by SMillerDev** [comment]
It's failing on style but I have no idea what "C: 12: 14: Use hash rockets syntax." means

File: Formula/elasticsearch@2.4.rb [hunk 1]
```diff
@@ -9,6 +9,8 @@ class ElasticsearchAT24 < Formula
 
   keg_only :versioned_formula
 
+  deprecate! date: "Februari 28, 2018"
```
Comment (MikeMcQuaid) [inline Formula/elasticsearch@2.4.rb:12]:
```suggestion
  deprecate! :date => "February 28, 2018"
```

(and same below). `brew style --fix homebrew/core` will fix these for you (but not the `February` typo)

File: Formula/go@1.12.rb [hunk 2]
```diff
@@ -13,6 +13,8 @@ class GoAT112 < Formula
 
   keg_only :versioned_formula
 
+  deprecate! date: "Februari 25, 2020"
```
Comment (MikeMcQuaid) [inline Formula/go@1.12.rb:16]:
```suggestion
  deprecate! :date => "February 25, 2020"
```

**Comment by SMillerDev** [comment]
Only failing on the changed kops hash now.

**Comment by MikeMcQuaid** [comment]
@smillerdev Could you update that as part of this?

**Comment by SMillerDev** [comment]
The hash is changed but the version isn't, that's what it's failing on

**Comment by MikeMcQuaid** [comment]
> The hash is changed but the version isn't, that's what it's failing on

@SMillerDev Yeh, @Bo98 confirmed above it's not a problem so would be good to update if possible.

**Comment by SMillerDev** [comment]
> @SMillerDev Yeh, @Bo98 confirmed above it's not a problem so would be good to update if possible.

It is, it's an audit failure

**Comment by MikeMcQuaid** [comment]
> It is, it's an audit failure

Can you fix my brain instead? :sob:

✅

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by reitermarkus** [comment]
I'm not sure about the `deprecate! :date =>` naming. How about just

```ruby
deprecate_on "February 29, 2020"
```

?

**Comment by Bo98** [comment]
I think the idea is that the date part is optional.

```rb
deprecate!
```

is valid.

**Comment by bayandin** [comment]
We could make it `deprecate! :on => "February 28, 2018"`, maybe it looks better

**Comment by MikeMcQuaid** [comment]
To be clear: the DSL is already implemented.

**Comment by reitermarkus** [comment]
> To be clear: the DSL is already implemented.

Well, it wasn't used anywhere yet, was it?

**Comment by MikeMcQuaid** [comment]
> Well, it wasn't used anywhere yet, was it?

It was documented and released so: who knows.

**Comment by sjackman** [comment]
Can we please, pretty please, with cherries on top, use the ISO-8601 date format `2018-02-28`? I am prepared to die on this hill. 😁 https://en.wikipedia.org/wiki/ISO_8601

**Comment by SMillerDev** [comment]
date format xkcd

Can you make an audit for it @sjackman?

**Comment by sjackman** [comment]
Good idea, Sean. Yes, can do. unless you want to poach it, in which case, 💞

**Comment by MikeMcQuaid** [comment]
I'm fine with a date format being recommended if it's audited.

**Pull request closed without merging**